Free Trials: Show upgrade notice on domain search and in cart during trial

drewblaisdell opened the pull request:
Fixes #1119.

![screen shot 2015-12-30 at 5 11 57 pm](https://cloud.githubusercontent.com/assets/1130674/12059902/5ee268ba-af19-11e5-913e-98d451161b37.png)

This PR:
- Adds a notice above `DomainSearch` when the user has a free trial.
- Adds a message in `PopoverCart` when the user has a free trial and a domain in their cart.

**Testing**
- Assert that the top notice in the screenshot is displayed to sites with a free trial on `DomainSearch`.
- Assert that the message in the cart in the screenshot is displayed in `PopoverCart` to sites with a free trial that have a single domain in their cart.

- [x] Product review
- [x] Code review

stephanethomas commented on client/my-sites/upgrades/cart/cart-trial-ad.jsx:
```diff
@@ -0,0 +1,48 @@
+	if ( isDataLoading ||
+		! currentPlan.freeTrial ||
+		cartItems.getDomainRegistrations( cart ).length !== 1 ) {
+		return <noscript />;
```
Just curious, why did you return `<noscript />` instead of `null`?

drewblaisdell replied:
Stateless components can't return `null`, and React uses `<noscript />` when stateful components that return `null`. I updated the first commit with a comment about this behavior and a link to the relevant GitHub issue.

stephanethomas commented on client/my-sites/upgrades/domain-search/domain-search.jsx:
```diff
@@ -10,12 +11,15 @@ var page = require( 'page' ),
 
-module.exports = React.createClass( {
+var DomainSearch = React.createClass( {
 	displayName: 'DomainSearch',
```
Do we still need to define `displayName` since we assign the component to a variable?

stephanethomas commented:
Tested and works fine.

I have two remarks though. First one is that it's still possible to use domain credits without upgrading but I guess this is something that will be fixed in another pull request. Second one is about the `day 0` mention in the notice - I would have expected it to be `day 1` instead:
> You are currently on day 0 of your WordPress.com Premium trial. Get this domain for free when you upgrade. Upgrade Now

stephanethomas commented:
Just noticed a small glitch - the notice is still visible even if the shopping cart already contains a plan:

![screenshot](https://cloud.githubusercontent.com/assets/594356/12095184/4ca2f9e2-b30d-11e5-9724-37f085b82c72.png)

drewblaisdell commented:
Thanks for looking at this, @stephanethomas. I made some updates by changing/adding commits:

- Added a comment for the use of `<noscript />`.
- Removed the `displayName` property of `DomainSearch`.
- Changed `getDaysSinceTrialStarted` to `getDayOfTrial` to indicate that the first day of a trial is "day 1".
- Updated `CartTrialAd` to not display the ad if the cart contains a plan.

> First one is that it's still possible to use domain credits without upgrading but I guess this is something that will be fixed in another pull request.

This is tracked with #1522.

spncrb commented:
Works as expected, minus caveats mentioned by Stéphane :+1:

drewblaisdell commented:
Thanks @spncrb.

In order to avoid calling .bind() inside a component's render(), I added a commit to change to the standard component syntax in `CartTrialAd` and `FreeTrialNotice`. This is a pretty small change, but I'm going to put this back into code review for a final :+1:

stephanethomas commented:
Code looks good. Let's ship it!
